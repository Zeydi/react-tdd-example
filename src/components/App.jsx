import React, { Component } from "react";
import { Button } from "react-bootstrap";
import Gift from "./Gift";

class App extends Component {
  constructor() {
    super();
    this.state = {
      giftList: [],
    };
  }

  addGift = () => {
    const { giftList } = this.state;
    const ids = giftList.map((gift) => gift.id);
    const max_id = giftList.length > 0 ? Math.max(...ids) : 0;

    giftList.push({ id: max_id + 1 });
    this.setState({ giftList });
  };

  removeGift = (id) => {
    const giftList = this.state.giftList.filter((gift) => gift.id !== id);
    this.setState({ giftList });
  };
  render() {
    const { giftList } = this.state;
    return (
      <div>
        <h2>Gift Giver</h2>
        <div className="gift-list">
          {giftList.map((gift) => {
            return (
              <Gift key={gift.id} removeGift={this.removeGift} gift={gift} />
            );
          })}
        </div>
        <Button className="btn-add" onClick={this.addGift}>
          Add Gift
        </Button>
      </div>
    );
  }
}

export default App;
