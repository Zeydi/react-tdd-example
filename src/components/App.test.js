import React from "react";
import { shallow } from "enzyme";
import App from "./App";

describe("App", () => {
  const app = shallow(<App />);

  it("should renders coreectly", () => {
    expect(app).toMatchSnapshot();
  });

  it("should initializes state with empty list", () => {
    expect(app.state().giftList).toEqual([]);
  });

  describe("when clicking add gift button", () => {
    const id = 1;
    beforeEach(() => {
      app.find(".btn-add").simulate("click");
    });

    afterEach(() => {
      app.setState({ giftList: [] });
    });

    it("should add  gift to state", () => {
      expect(app.state().giftList).toEqual([{ id }]);
    });

    it("should creates a Gift component", () => {
      expect(app.find("Gift").exists()).toBe(true);
    });

    describe("when clicking remove gift button", () => {
      beforeEach(() => {
        app.instance().removeGift({ id });
      });

      afterEach(() => {
        app.setState({ giftList: [] });
      });

      it("should add  gift to state", () => {
        expect(app.state().giftList).toEqual([{ id: 1 }]);
      });

      it("should creates a Gift component", () => {
        expect(app.find("Gift").exists()).toBe(true);
      });
    });
  });
});
