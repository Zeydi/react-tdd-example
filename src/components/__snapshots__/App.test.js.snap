// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should renders coreectly 1`] = `ShallowWrapper {}`;
